keps/draft-kuttl-as-a-library.md

```diff
@@ -0,0 +1,193 @@
+---
+kep-number: 0
+short-desc: Overview of providing KUTTL for use as a Go library
+title: KUTTL as a Library
+authors:
+  - "@mpereira"
+  - "@jbarrick-mesosphere"
+owners:
+  - "@mpereira"
+  - "@jbarrick-mesosphere"
+editor: TBD
+creation-date: 2020-02-13
+last-updated: 2020-02-13
+status: provisional
+see-also:
+  - KEP-2
+---
+
+# KUTTL as a Library
+
+## Table of Contents
+
+* [Summary](#summary)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+
+
+## Summary
+
+Currently, the test harness supports only writing tests in YAML which works well for the majority of cases but there are some use-cases that are currently hard to support with the fully declarative tests (for example, asserts that require context from multiple fields or objects or fuzzy asserts). Additionally, there are some developers who would prefer to write tests in Go due to their personal familiarity with the language. There are also other developers that would prefer to write fully declarative tests and not need to deal with a full on programming language.
+
+Rather than building an entirely new library, we will unify the library and test harness efforts to ensure that improvements to one benefit both and that we do not duplicate work or fragment the community.
+
+### Goals
+
+* Support writing tests in Go for developers that prefer a traditional programming approach.
+* Increase code reuse, prevent the need for each operator to have their own special testing library.
+* Share expertise and code between fully declarative and Go-based tests.
+* Allow for "breaking the glass" when declarative tests are not expressive enough.
+
+### Non-Goals
+
+* Creating a new spin-off testing project.
+* Fragmentation of our operator developer community.
+* Dropping support for YAML-based declarative tests.
+
+### Principles
+
+* Useful logging by default
+* Depend on existing libraries whenever possible
+* Retrying by default where it makes sense
+* Ease of adding retrying to non-retried functions
+* It is better to be consistent than to be "perfect"
+* Err on the side of flexibility
+* Err on the side of returning errors instead of panicking/Fatalf/etc.
+* Cohesive logging output (parameters, formatting, etc.)
+* Separation of concerns between domains (KUDO, Kubernetes, etc.)
+
+## Proposal
+
+We will improve and break-down the interface of the kuttl library to make it easier to consume as a downstream developer. Libraries that need to be broken out of kuttl:
+
+* A generic Kubernetes client (currently this is contained inside of `pkg/test/harness.go`).
+* Kubernetes object constructors (this is already pretty good in `pkg/test/utils/kubernetes.go`).
+* Assertion methods (needs better documentation for library usage).
+* Test setup and tear down methods (is currently coupled to steps).
```

Comment:
Wouldn't this, as well as assertions, be something other libraries can provide? Test frameworks like [ginkgo](https://github.com/onsi/ginkgo) provide these functions among others. There's also a lot of assertion libraries, e.g. [testify](https://github.com/stretchr/testify).
Though it would make sense to provide assertion helpers for more complex test scenarios, e.g., asserting that a plan completed `assertPlanCompleted(instance.plan["deploy"])`.

Reply:
I'll expand on what I meant in the doc in a bit - but:

* test setup and tear down methods: this would be more along the lines of the namespace creation and deletion methods below - kubernetes-specific test setup. Not about enforcing a test framework / structure or even necessarily a way to run them.
* assertions: what I meant here is more complex assertions: generally kubernetes assertions like you mentioned, but also supporting the same type of asserts we do in the test harness CLI (https://kudo.dev/docs/testing/asserts-errors.html). So, for example, you could use Go to generate a bunch of objects, apply them, and then you can use the assertion method to wait for your expected state instead of having to roll your own thing.